### PR TITLE
Ensure every instance of StreamFilter and PassThroughFilter to contain only one copy of StreamFilterBase.

### PR DESCRIPTION
**Pull request by wu-bin**
Change Stream(Decode|Encode)Filter to inherit from StreamFilterBase virtually.
Move the default implementation of PassThrough(Decoder|Encode)Filter::onDestroy into StreamFilterBase.

Commit Message: Ensure every instance of StreamFilter and PassThroughFilter to contain only one copy of Stream(Decode|Encode)Filter and StreamFilterBase.
Additional Description:
Risk Level: Medium
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: envoy/http/filter.h [hunk 1]
```diff
@@ -857,7 +857,7 @@ class StreamFilterBase {
    * onDestroy() invoked. Filters that cross-register as access log handlers receive log() before
    * onDestroy().
    */
-  virtual void onDestroy() PURE;
+  virtual void onDestroy() {}
```
Comment (yanavlasov) [inline envoy/http/filter.h:860]: Can you keep this method pure virtual if you add override onDestroy in the PassThroughFilter? Or is it causing more build errors than that?

**Review by yanavlasov — COMMENTED** [review]
/wait-any

File: envoy/http/filter.h (hunk 1 above)
Comment (wu-bin) [reply envoy/http/filter.h:860]: Thanks Yan, but I think if we do that, an existing class will not compile if
- It derives directly from `PassThroughDecoderFilter` or `PassThroughEncoderFilter`, and
- It doesn't override onDestroy, i.e. it relies on the empty implementation in `PassThroughDecoderFilter` or `PassThroughEncoderFilter`.

Comment (wu-bin) [reply envoy/http/filter.h:860]: /assign @yanavlasov

Comment (yanavlasov) [reply envoy/http/filter.h:860]: Did you try it? Can you post the compiler error, please?

Comment (wu-bin) [reply envoy/http/filter.h]: As discussed, I change to have 3 `onDestroy` implementations in `PassThroughDecoderFilter`, `PassThroughEncoderFilter`, and `PassThroughFilter`. PTAL.

**Review by yanavlasov — APPROVED** [review]

**Pull request merged**